### ISSUE TITLE
Counting the number of collaborations using sparse matrix

### ISSUE THREAD

**Issue by hosseinfani** [fani-lab/OpeNTF#117]
We want to know which experts have collaborated with each other the most, using the sparse matrix representation of teams.

**Comment by VaghehDashti** [comment]
As we discussed before, multiplying the authors per team matrix to its transpose will result in a matrix containing the number of collaborations of each author with other authors in each row and the diagonal of the matrix is the number of papers of each author. The same procedure can be done for skill per team matrix.
[number of authors, number of papers]*[number of papers, number of authors] = [number of authors, number of authors]
[1 0 0 1 0]    [1 1 0 0]    [2 2 1 1]
[1 0 1 1 0] * [0 0 1 1]=  [2 3 2 1]
[0 1 1 1 1]    [0 1 1 0]    [1 2 4 3]
[0 1 0 1 1]    [1 1 1 1]    [1 1 3 3]
                    [0 0 1 1]

**Comment by hosseinfani** [comment]
@VaghehDashti ya, seems this is correct. we have to put all these with proofs in another short paper :)

#82

**Comment by hosseinfani** [comment]
@VaghehDashti but this only gives the pairwise collaboration. How about finding the number of times that three (four, five, ...) members together collaborated in teams?

**Comment by hosseinfani** [comment]
@gabrielrueda 
Any update?

**Comment by gabrielrueda** [comment]
@hosseinfani  Yes, I have an update:
I looked into what was said in this issue about how the matrix times it's transpose will give the pairwise collaborations for every combination of members. 
I have created a function does the multiplication and puts the data into a sparse matrix:

```Python
def getPairwiseCollaborations(x_m, y_m):
    rows = []
    columns = []
    data = []
    for i in range(0,13):
        row_x = (x_m.getrow(i).asformat("array"))[0]
        for j in range(0,13):
            col_y = (y_m.getcol(j).asformat("array"))
            val = 0
            for k in range(0,30):
                val += row_x[k] * col_y[k][0]
            if(val != 0):
                data.append(val)
                rows.append(i)
                columns.append(j)
    return sci.sparse.coo_matrix((data, (rows, columns)))  
 ```
I have looked a bit into finding the number of times 3, 4, 5, ... members collaborate at once, but haven't come up with anything yet.

**Comment by hosseinfani** [comment]
@gabrielrueda you don't need to implement the multiplication. Why not using the numpy's `transpose` and `matmul` functions??
Also, please test your function on the test example above and show the result here.

**Comment by hosseinfani** [comment]
@gabrielrueda you can write and test your code in a google colab. then share the script with me here. 
Also, let's first test it for the sample by @VaghehDashti above. If everything is alright, we can run it on the teamsvecs.pkl.
thanks.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
I figured out how to make it work with `tranpose` and `matmul` functions as shown in the `getCollabsVer2()` function. I also found out that you can use the the `@` operator for matrix multiplication as shown in the `getCollabsVer3()` function.

I tested the code on Google Colab with the sample above and it worked. I have attached the code and output below:

```Python
import pickle

import numpy as np
import scipy.sparse as sci

# Using the matmul() function from numpy library
def getCollabsVer2(m_matrix):
    return sci.sparse.coo_matrix(np.matmul(m_matrix.transpose().toarray(), m_matrix.toarray()))

# Using "@" operator -> does Matrix Multiplication
def getCollabsVer3(m_matrix):
    return m_matrix.transpose() @ m_matrix


def main():  
  # Sparse Matrix teamsvecs.pkl
  # with open('data/preprocessed/dblp/toy.dblp.v12.json/teamsvecs.pkl', 'rb') as tfile: 
  #     matrix=pickle.load(tfile)

  # Matrix from @VaghehDashti:
  A = sci.coo_matrix([[1,1,0,0], [0,0,1,1], [0,1,1,0], [1,1,1,1], [0,0,1,1]])
  
  outMat = getCollabsVer3(A)
  # outMat = getCollabsVer2(A)

  # print(outMat)

  f = open("output.txt","wt")

  # Prints the output onto output.txt
  for i in range(0,4):
      f.write("[")
      out = (outMat.getrow(i).asformat("array"))[0]
      for j in range(0,len(out)):
          f.write(str(out[j]) + " ")
      f.write("]\n")


main()
 ```

OUTPUT:
![image](https://user-images.githubusercontent.com/35504507/169416924-b50a38bb-9209-4d9d-a24d-7f2123f19497.png)

**Comment by hosseinfani** [comment]
@gabrielrueda Thank you. Now, we need to order the pairs of authors descending and print out the top-k (eg, top-2). 

You only need to consider the upper (or lower) part of the final matrix, given the row i and column j is the indexes of author i and author j.

In our example, it is

**Comment by hosseinfani** [comment]
@gabrielrueda btw, thank you for sharing the amazing @ operator on matrices :)

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
Here is another update. I have made an implementation of getting the top-k pairs of authors. 
However I wasen't sure...
1. if I should remove the diagonal entries before finding the Top-K pairs (since author i = author j)
2. how I will deal with duplicates when printing out the top-k (e.g. Suppse the order was 4,3,3 for the pairs were (1,2), (5,1), (3,2) respectively. If it were top-2 would I include the pair (5,1) or (3,2))

Here is the function I wrote so far:
```Python
def getTopK(m_matrix, k):
    result = [0] * k
    upperTri = sci.triu(m_matrix)
    
    
    indexes = sci.find(upperTri)

    # Sorts the List of Elements in Upper Triangular
    for i in range(0, len(indexes[0]) -1):
        max = i
        for j in range(i+1, len(indexes[0])):
            if(indexes[2][j] > indexes[2][max]):
                max = j
        # Swap the "i" (row) coordinate:
        temp = indexes[0][i]
        indexes[0][i] = indexes[0][max]
        indexes[0][max] = temp
 
        # Swap the "j" (column) coordinate:
        temp = indexes[1][i]
        indexes[1][i] = indexes[1][max]
        indexes[1][max] = temp

        # Swap the values
        temp = indexes[2][i]
        indexes[2][i] = indexes[2][max]
        indexes[2][max] = temp
    

    rows = indexes[0][0:k]
    columns = indexes[1][0:k]
    data = indexes[2][0:k]


    return sci.coo_matrix((data, (rows, columns)),shape=(4,4))
```

This is the output for k=4:
![image](https://user-images.githubusercontent.com/35504507/169926675-854b2477-4790-4fec-bcd2-187725b26609.png)

**Comment by hosseinfani** [comment]
@gabrielrueda 
Thank you for the update.

1. Yes, you should not consider the diagonal. Either you can filter them from the beginning or filter them at the end (remove those with i and j equal from the final list). Up to you. I prefer the first one.
2. You're right but you don't need to do anything. The [:k] cut-off you did is correct.

Also, the output should not be shown as a matrix but a list as I showed above. You did right for coo_matrix. We can simply print coo_matrix content.

Also, remove the shape=(4,4) as the function may be called for a matrix of any size.

Please update me and then we can move to the next part which is finding the collaboration of three people, four people or in general n people together.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
I have replaced `shape=(4,4)` with `shape=upperTri.get_shape()` and filtered out the diagonal entries using `upperTri.setdiag(0)`.

Here is the output for k=4:
![image](https://user-images.githubusercontent.com/35504507/170546146-b756c78c-ac94-4a70-badd-09fc1a957d46.png)

Here is the function:
```Python
def getTopK(m_matrix, k):
    result = [0] * k
    upperTri = sci.triu(m_matrix)

    upperTri.setdiag(0)

    indexes = sci.find(upperTri)

    # Sorts the List of Elements in Upper Triangular
    for i in range(0, len(indexes[0]) -1):
        max = i
        for j in range(i+1, len(indexes[0])):
            if(indexes[2][j] > indexes[2][max]):
                max = j
        # Swap the "i" (row) coordinate:
        temp = indexes[0][i]
        indexes[0][i] = indexes[0][max]
        indexes[0][max] = temp
 
        # Swap the "j" (column) coordinate:
        temp = indexes[1][i]
        indexes[1][i] = indexes[1][max]
        indexes[1][max] = temp

        # Swap the values
        temp = indexes[2][i]
        indexes[2][i] = indexes[2][max]
        indexes[2][max] = temp
    

    rows = indexes[0][0:k]
    columns = indexes[1][0:k]
    data = indexes[2][0:k]


    return sci.coo_matrix((data, (rows, columns)),shape=upperTri.get_shape())
 ```

**Comment by hosseinfani** [comment]
@gabrielrueda Thank you.
Before we move on, would you please render the results of getTopK() in a histogram? Something like this:
![image](https://user-images.githubusercontent.com/8619934/170549397-76273b3e-8116-400a-9850-d39e5d28cda6.png)

**Comment by gabrielrueda** [comment]
@hosseinfani 
I have rendered the results using the following code:

```Python
def plotTopK(result):
    # indices[0] for x coordinate
    # indices[1] for y coordinate
    indices = result.nonzero()

    data = result.data
    k = len(data) # Gets the k value

    fig = plt.figure()
    ax = fig.add_axes([0.1,0.1,0.7,0.7])
    
    # Form X-Axis Categories:
    xAxis = []
    for i in range(0,k):
        xAxis.append("(" + str(indices[0][i]) + "," + str(indices[1][i]) + ")")
    
    # Generates the Graph:
    ax.bar(xAxis,height=data)
    ax.set_title("Top " + str(k) + " Collaborations")
    ax.set_ylabel("Frequency")
    ax.set_xlabel("Pairs in Order of Top-" + str(k))

    plt.show()
 ```
The output is: 
![image](https://user-images.githubusercontent.com/35504507/170578330-472913ae-2b72-4d21-9aff-2341d059ec37.png)

**Comment by hosseinfani** [comment]
@gabrielrueda thank you. this is awesome. 
I'll explain the next step tomorrow.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Congrats! You're making your first official contribution to the OpeNTF project.
Please put all your final code in a file colabstat.py and make a pull request to our main branch. 
Thank you.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Now, let's find out how to do stats on experts that have more than two collaborations. For instance, find out the top-k experts who have 3 collaborations together, 4 collaborations, ... and in general n collaborations.

The naive way would be to count them. For instance, if n=3, we can find all triple experts ([we will have combination of 3 out of all experts](https://en.wikipedia.org/wiki/Combination)) then count all teams that have each triple.

Example: N={m1,m2,m3,m4} we will have combination of 3 out of 4 = C(3,4) = 4
{m1,m2,m3} => count the teams that have them all 
{m1,m4,m3} => count the teams that have them all 
{m1,m2,m4} => count the teams that have them all 
{m2,m3,m4} => count the teams that have them all 

But we need to solve it using matrix multiplication or better/faster algorithms.
Please do a search and try to find or devise an algorithm to solve this task.

**Comment by gabrielrueda** [comment]
Hi, @hosseinfani 
Here is an update:
I have decided to start my trying to find 3 collaborations together. I tried a few things with the matrix from the pairwise collaborations and observed the following:

For this first matrix, there is no team of 3 collaborators:
![image](https://user-images.githubusercontent.com/35504507/171214157-6e01aad3-5f69-4972-88bb-64a8885b8dbf.png)
- The top matrix is the original
- The bottom matrix is itself * it's inverse
- The Diagonal entry = sum of the elements in it's row (excluding diagonal)
    - So... 3 = 1 + 1 + 1, 2 = 1 + 1 + 0 ...
    - This means no teams

For this second matrix, there is 1 team of 3 collaborators:
![image](https://user-images.githubusercontent.com/35504507/171214581-568a4640-9cc8-4824-8f2f-daac9b93e4dc.png)
- Top is original, bottom is itself * inverse
- For each row, find "sum of the entries in it's row (excluding diagonal) - diagonal entry"
- Row 1 (m_1) = (2+1+1) - 3 = 4-3 = 1
- Row 2 (m_2) = 4 - 3 = 1
- Row 3 (m_3) = 3 -3 = 0
- Row 4 (m_4) = 3 -2 = 1
- This means we have 1 from m_1, 1 from m_2, and 1 from m_4 -> so the team is m_1, m_2 and m_4

For this third matrix, there are 2 teams of 3 collaborators:
 ![image](https://user-images.githubusercontent.com/35504507/171215742-d15c6e82-e174-4f8c-9550-699f104ecdfc.png)
- Since we have 2 from m_1, 2 from m_2, 1 from m_3, and 1 from m_4...
    - We have the following collaborations:
    1. m_1, m_2, m_4
    2. m_1, m_2, m_3
 
However, in this fourth matrix, where there is 1 team of 4 collaborators, I ran into problems counting the collaborations:
![image](https://user-images.githubusercontent.com/35504507/171216453-bd64e51b-9334-46bd-b408-056b09932dfa.png)
- I tried to make the following collaborations:
1. m_1, m_2, m_3
2. m_1, m_2, m_4,
3. m _1, m_3, m_4
-> But now we have m_2 and m_3 left with an incorrect number of collaborations

Now, I will continue working with this idea and see if I figure out the last scenerio

**Comment by hosseinfani** [comment]
@gabrielrueda 
thank you for the update. 
I was thinking of a counter-example to counting the triads the way I said. What if m1,m2 are in t1, m2,m3 in t2, and m3,m1 in t3, then they form a triad but they are not part of a single team altogether. right?

**Comment by gabrielrueda** [comment]
@hosseinfani 
Yes, your right. I will try to modify the function or the matrix beforehand to detect this special case.

**Comment by hosseinfani** [comment]
@gabrielrueda I think this doesn't work for the general case. I'm asking a friend of mine who's mathematician. Will uodate you here.

This is a combinatorial problem. 

For now,please implement the naive method, that is counting them by enumerating all possible cases, as I explained above. 

Meanwhile, you can research more about it to see if there is a better algorithm.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
I have just implemented the naive method of counting all possible cases using combinations. I will now research more to see if I can find a better algorithm. Here is the code I wrote:

```Python
# Combinations Method
def threeWayCollabs_combinations(m_matrix):         
    rowIndexes = []
    for i in range(0, m_matrix.shape[1]):
        rowIndexes.append(i)

    # comb is a list of all the possible combinations
    # comb (if there are 4 rows) = [0,1,2], [0,1,3], ...
    comb = combinations(rowIndexes, 3) 

    # Will record the count for each combination
    # The index for count aligns with it's respective combination
    count = []
    for testCase in list(comb):
        tempCount = 0
        for i in range(0, m_matrix.shape[0]):
            row = (m_matrix.getrow(i).asformat("array"))[0]
            if (row[testCase[0]] == 1 and row[testCase[1]] == 1 and row[testCase[2]] == 1):
                tempCount += 1
        count.append(tempCount)

        # Prints the results:
        if tempCount != 0:
            print(str(testCase) + " => " + str(tempCount))
 ```

Here is the output:
![image](https://user-images.githubusercontent.com/35504507/173125669-d153fbe8-3d30-4847-8a34-3294848dc1c5.png)

**Comment by hosseinfani** [comment]
@gabrielrueda 
Thank you. 
Please make it n-way. Something like this:
```
def nwayColabs(n, k): #n: number of members in collaboration, k: top-k most 
   if n ==2: 
       m = getColabsVer3()
       getTopK(m, k)
   else:
       #your new nwaycolabs
       #selecting the top-k
```
Also, we want to reuse `plotTopK` to plot the top-k.

**Comment by hosseinfani** [comment]
@gabrielrueda 
@mahdis-saeedi was able to find an answer for counting n-way collaboration. Basically, if a member `m_i` has a vector of membership to teams `v_m_i`, then the count of collaborations between any n members is the `dot product` of their membership vectors. For instance, for 3-way, for any m_i, m_j, m_k => colabs = v_m_i  [dot] v_m_j [dot] v_m_k

Indeed, our matrix multiplication for 2-way was the special case of this. When we do matrix multiplication, we are doing dot product of rows in first matrix to the columns to the second matrix. 

From math, dot product of two vectors is the element-wise multiplication followed by a sum:

v_1 = [a,b,c,d]
v_2 = [A, B, C, D]

v_1 [dot] v_2 = aA + bB + cC + dD

She will post her solution here, and we can discuss it further here.

**Comment by mahdis-saeedi** [comment]
@hosseinfani 
@gabrielrueda 
[counting collaborations.pdf](https://github.com/fani-lab/OpeNTF/files/8884588/counting.collaborations.pdf)

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
Here is an update. I first made the nwayCollabs() function like you said for the naive method for the combinations (which is copied below). Now, I am working on implementing the solution provided by `@mahdis-saeedi`. However I ran into a problem:

Suppose you have matrix:
[ 1 0 0 0 1 ]
[ 0 0 1 1 1 ]
[ 1 1 0 0 0 ]

So I would count the collaborations between m1, m2 and m3 like this
= v_m_1 [dot] v_m_2 [dot] v_m_3
= [1 0 1] [dot] [0 0 1] [dot] [0 1 0]
= [1]  [dot] [0 1 0]
= [0 1 0]

The output is a vector, but I believe it should be a scalar value. I was wondering if you could let me know if I am calculating the dot product of these three vectors in the wrong way, or if the output is supposed to be vector. 

Finally, below I will attach my functions for the naive way (w/ Combinations):
```Python
def getTopK_nWays(collabs, k):
    for i in range(0, len(collabs) - 1):
        max = i
        for j in range(i+1, len(collabs)):
            if(collabs[j][1] > collabs[max][1]):
                max = j
        temp = collabs[i]
        collabs[i] = collabs[max]
        collabs[max] = temp
    print(collabs)
    return collabs[0:k]


# Format: [coordinate, number]
def nCollabs_combinations(m_matrix, n):      
    rowIndexes = []
    for i in range(0, m_matrix.shape[1]):
        rowIndexes.append(i)
    

    # comb is a list of all the possible combinations
    # comb (if there are 4 rows) = [0,1,2], [0,1,3], ...
    comb = combinations(rowIndexes, n) 
    # print(list(comb))

    # Will record the count for each combination
    # The index for count aligns with it's respective combination
    collabs = []
    for testCase in list(comb):
        tempCount = 0
        for i in range(0, m_matrix.shape[0]):
            row = (m_matrix.getrow(i).asformat("array"))[0]
            for j in range(0, n):
                if(row[testCase[j]] != 1):
                    tempCount -= 1
                    break
            tempCount += 1

        if tempCount != 0:
            collabs.append([testCase, tempCount])

    # Prints the results:
    return collabs


def nwayCollabs(m_matrix, n,k):
    if n==2:
        m = getPairwiseCollabs(m_matrix)
        getTopK(m, k)
    else:
        m = nCollabs_combinations(m_matrix, n)
        topK = getTopK_nWays(m, k)
        plotTopK_new(topK)

```

**Comment by hosseinfani** [comment]
Hi @gabrielrueda 
Sorry for the late reply.
You're right. The dot product of two vectors is a value that cannot be further dot product to a third or more vector.

We can define dot product of more than two vectors as the element-wise multiplication of the vectors and then sum all. For instance, if we have three vectors (members) and 5 teams:

[ 1 0 0 0 1 ]
[ 0 0 1 1 1 ]
[ 1 1 0 0 0 ]
==
1st col:  1*0*1 +
2nd col: 0*0*1 + 
3rd col:  0*1*0 +
4th col:  0*1*0 +
5th col:  1*1*0 +
==
0 meaning that these three people never worked all together

**Comment by hosseinfani** [comment]
@gabrielrueda 
Also, for the updated code, please send another pr on the same file so I can merge and we keep track of changes. Thank you.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani,
I wrote the function using the element-wise multiplication and it worked. The only thing is that I used combinations to get the rows (vectors) I would do the element-wise multiplication on. Should I try to find a way without listing the combinations, or just leave it as is. 

Here are the functions:
```Python
def nwayCollabs_2(m_matrix, n,k):
    m = nCollabs_vectors(m_matrix, n)
    plotTopK_nWays(getTopK_nWays(m, k))


def nCollabs_vectors(m_matrix, n):
    collabs = []
    rowIndexes = []
    for i in range(0, m_matrix.shape[1]):
        rowIndexes.append(i)

    # comb is a list of all the possible combinations
    # comb (if there are 4 rows) = [0,1,2], [0,1,3], ...
    comb = combinations(rowIndexes, n) 

    for team in list(comb):
        dotP = (m_matrix.transpose().getrow(team[0]).toarray())[0]
        for i in range(1, n):
            dotP = dotP * (m_matrix.transpose().getrow(team[i]).toarray())[0]
        collabs.append([team, np.sum(dotP)])
    
    return collabs
```
And here is the result for matrix (n=3, k=2):
[[1 1 0 0]
 [0 0 1 1]
 [1 1 0 1]
 [1 1 0 1]
 [1 0 1 1]]

![image](https://user-images.githubusercontent.com/35504507/174632124-3a90e2ef-d24f-4fea-950d-36bc1cbc6d5f.png)

**Comment by hosseinfani** [comment]
Hi Gabriel,
@gabrielrueda 

I like this new version more.

I merged your previous code and did some changes. Please update your code and apply your new changes. Send the PR then. 

One change I make the x axis ticks to be vertical (in case we want to render the names (if available)). But, the plot figure does not fully show them. Please fix that too.

After your PR, we can safely close this task and start the new one.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Thank you very much for your clean and readable codelines. It's my pleasure reviewing your codes :)

**Comment by hosseinfani** [comment]
@gabrielrueda 
I tried to run the code on the whole dataset. 
- First, we need to make the lines of code as optimized as possible. 
- Second, we need to make it parallel. 

#Publications (teams) | 4,877,383
#Unique Authors (members) | 5,022,955

**Comment by hosseinfani** [comment]
@gabrielrueda 
are you available tomorrow or wednesday to talk about the new task? preferably in-person on campus. This new task is a little bit challenging ...
Please let me know.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Hope all is well with you.
I ran the code on the entire dataset and also on the filtered version of it. 

Unfortunately, after 4-days I had to stop it. I know that the time complexity is O(n^2) and n~=14,000 in the filtered version of dataset. 

I ran it for {2,3,4,5}-way collabs in parallel on the latest code at [here](https://github.com/fani-lab/OpeNTF/blob/main/src/misc/collabs.py)

We need to optimize the code.

**Comment by gabrielrueda** [comment]
Hello @hosseinfani, 
I took a look at the code and found a few adjustments that might optimize the code more:
1. I transposed the matrix before the loop, so it doesn't do the transpose action in every iteration. In function: `getnwayCollabs()`
2. I filtered out all the groups with no collaborations before returning the collabs list in `getnwayCollabs()`
3. I also changed my bubble sort to use the sort() function built-in to python. From my understanding, the built-in function uses an adaptive mergesort and runs at O(n*log(n))

I made the changes above to the code and created a pull request for it.

In addition, I had an another idea to optimize the sorting, but I am not sure if it will do anything to optimize. 
Here's the idea:
1. All of the 1's from the collabs list will be placed another list
2. The collabs list will sort without any of the 1's.
3. The list with the 1's will get attached to the end of the collabs list.

**Comment by hosseinfani** [comment]
@gabrielrueda 
Thank you. I will merge the pr.

I'm not sure I understood the second part (sorting idea). Can you explain with an example?

**Comment by gabrielrueda** [comment]
@hosseinfani 
Yes here is an example: 
collabs= { {(0,1,2), 1},  {(0,1,3), 3}, {(0,2,3), 1}, {(1,2,3), 2}}

1. STEP 1: Filter out the 1's and place into a new list
- collabsWithOne = { {(0,1,2), 1},  {(0,2,3), 1} }
- collabs = { {(0,1,3), 3}, {(1,2,3), 2}}

2. STEP 2: Sort the collabs List
- collabs = { {(0,1,3), 3},  {(1,2,3), 2}}

3. STEP 3: Attach the collabsWithOne to the end of the collabs list
- collabs = { {(0,1,3), 3},  {(1,2,3), 2}, {(0,1,2), 1},  {(0,2,3), 1} }

**Comment by hosseinfani** [comment]
@gabrielrueda 

Great idea. Let's define a threshold (default value=1) and filter out those entries below the threshold. There is no need to add them later because these entries have no benefit even if we consider them (e.g., what's the benefit of considering 1 collabs)

**Comment by gabrielrueda** [comment]
@hosseinfani 
I created the PR for filtering the entries.

**Comment by hosseinfani** [comment]
@gabrielrueda thank you.

**Comment by hosseinfani** [comment]
@gabrielrueda 
![image](https://user-images.githubusercontent.com/8619934/177694685-a5c40150-8c45-4ebf-bb38-59d1085e9392.png)

**Comment by hosseinfani** [comment]
@gabrielrueda 
it's done for n=2:

```
#100%|██████████| 101011791/101011791 [21:21:26<00:00, 1313.78it/s]

import pickle
with open("./data/preprocessed/dblp/dblp.v12.json.filtered.mt75.ts3/collabs-2-top10.pkl", 'rb') as f: a = pickle.load(f)
with open('./data/preprocessed/dblp/dblp.v12.json.filtered.mt75.ts3/indexes.pkl', 'rb') as f: indexes = pickle.load(f)
names = indexes['i2c']
from src.misc.collabs import *
plotTopK_nWays(a, names=names, savefig="./data/preprocessed/dblp/dblp.v12.json.filtered.mt75.ts3/collabs-2-top10.png")
```
![image](https://user-images.githubusercontent.com/8619934/177906038-9e9aa173-4bf6-45a0-b4eb-2dfaad5aeefd.png)

I let it run for top-1000.

-------------------------------
For n>2, it's not doable. Since, for example, n=3, we have nCr(14214, 3) = 478526524564 combinations and the loop does ~3500 iter/sec. So, ((478526524564/3500)/3600)/24 = 1582 days

We can leave it for future since it's not our priority now.

**Comment by hosseinfani** [comment]
![image](https://user-images.githubusercontent.com/8619934/177906704-18483a84-c008-4102-ba57-2af17df78058.png)

**Comment by hosseinfani** [comment]
![image](https://user-images.githubusercontent.com/8619934/177935355-28058462-d313-4afa-9ea9-d31ad331ebd9.png)
